### DISCUSSION TITLE
Keybindings stop working if the transient prompt is enabled in zsh

### DISCUSSION THREAD

**Discussion by lewis-yeung** [JanDeDobbeleer/oh-my-posh#2617] [Troubleshoot] [answered]
### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

### What happened?

I'm sure that <kbd>Home</kbd> and <kbd>End</kbd> work correctly before the transient prompt is enabled in zsh. However, once it is enabled, the two keybindings stop working. Tested without other plugins. If I run `zle -N zle-line-init`, which disables the transient prompt, the keybindings will work again.

### Theme

(This happens with any theme configs, basically irrelevant.)

### What OS are you seeing the problem on?

Linux

### Which shell are you using?

zsh

### Log output

(N/A)

**Comment by JanDeDobbeleer** [comment]
@lewis-yeung not something specific to oh-my-posh it seems. Can you have a look at this [issue](https://github.com/ohmyzsh/ohmyzsh/issues/3061)?

**Comment by lewis-yeung** [comment, marked as answer]
@JanDeDobbeleer Thank you for mentioning that issue. I'm sure that [`_posh-zle-line-init`](https://github.com/JanDeDobbeleer/oh-my-posh/blob/9238f3adbf5da839e391af4542c62f23461eec82/src/shell/scripts/omp.zsh#L67-L92) function has something to do with it, however I don't know how to fix it. Anyway, I can just make the two keys work on my end by binding them via:

```zsh
bindkey "^[[F" end-of-line
bindkey "^[[H" beginning-of-line
```

**Comment by JanDeDobbeleer** [comment]
@lewis-yeung did some research, can you validate the changes in #2595? I have no end/home keys...

**Comment by lewis-yeung** [comment]
@JanDeDobbeleer I have tried your quick fix but the problem persists...

**Comment by hydezhao** [comment]
I believe this is related to the `application mode` vs `normal mode` (keybinds values of arrows change between these two mode for example `"^[[A"` vs `"^[OA"` ) while I didn't figure out how to properly fix.

**Comment by StanSvec** [comment]
It happens in my case too. [This discussion](https://github.com/zsh-users/zsh-history-substring-search/issues/70) may be related. I'm fixing it by adding these lines in `.zshrc` after oh-my-posh init:
```
bindkey '^[[A' up-line-or-search
bindkey '^[[B' down-line-or-search
```

**Reply by JanDeDobbeleer** [reply]
I wonder if we should add this after we enable transient prompt.

**Reply by StanSvec** [reply]
Unfortunately my knowledge of shells/unix is insufficient to provide any advice. But it sounds like a good idea. To correct myself to restore the exact previous behavior of up/down arrow keys this needs to be added in my case:
```
bindkey '^[[A' up-line-or-beginning-search
bindkey '^[[B' down-line-or-beginning-search
```

**Comment by michalzielanski** [comment]
The solution from version 12.4.0 (b174c27) was good. `Home`/`End` keys worked. However, in version 12.5.1 (a94ce52) they do not work again.
Zsh version: 5.8.1 (x86_64-apple-darwin21.0)

**Reply by lewis-yeung** [reply]
@JanDeDobbeleer Should we add these back?

```zsh
bindkey '^[[F' end-of-line
bindkey '^[[H' beginning-of-line
```

**Comment by refactorized** [comment]
don't know what this is all about, but I tried oh-my-posh today and it broke zsh, because every time I press up or down arrow (the actual arrow keys on an abbple macbook, running xterm-256color emulation) I get:
```
No such widget `up-line-or-beginning-search'
```

**Reply by JanDeDobbeleer** [reply]
@refactorized is your zsh version up-to-date?

**Reply by refactorized** [reply]
I thought it was reasonably up to date, 5.8.1 i think, but I just updated to `zsh 5.9 (x86_64-apple-darwin21.3.0)` and seem to be having the same issue.  Do I need to run any kind of completion scripts?  I do some compinit stuff later on in the rc, i don't know if that affects anything, but I only have problems if I eval the `oh-my-posh init zsh` stuff.

**Reply by JanDeDobbeleer** [reply]
@refactorized check #2999

**Reply by michalzielanski** [reply]
@JanDeDobbeleer The fix for someone is a regression for others. After the changes from e884a6a, [fuzzy find history](https://github.com/ohmyzsh/ohmyzsh/blob/master/lib/key-bindings.zsh#L34-L51) (Start typing + [Up/Down-Arrow]) does not work for me.
Maybe restoring keybindings should be done conditionally? Maybe like this:
```zsh
if [[ $(zle -la | grep ^up-line-or-beginning-search$) ]]; then
  bindkey '^[[A' up-line-or-beginning-search
  bindkey '^[[B' down-line-or-beginning-search
fi
```
